{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "ABIN_BLAST.ipynb",
      "provenance": [],
      "collapsed_sections": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    }
  },
  "cells": [
    {
      "cell_type": "code",
      "metadata": {
        "id": "rOBbbWRUEZxI"
      },
      "source": [
        "#Imports\n",
        "import math\n",
        "import time\n",
        "from tkinter import * \n",
        "from tkinter import ttk\n",
        "from tkinter.filedialog import askopenfile \n",
        "from PIL import ImageTk,Image"
      ],
      "execution_count": 1,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "AjSbyAIFEofY"
      },
      "source": [
        "class BLAST:\n",
        "  \"\"\"\n",
        "  Local Alignment of a query DNA Sequences with a database sequence\n",
        "  \"\"\"\n",
        "  def __init__(self, l_mer_length = 11 , query_string = \"\" , sequence_string = \"\", hssp_score = 8):\n",
        "    self.start_time = time.time()\n",
        "    query = query_string.upper().replace(\"N\" , \"\")\n",
        "    dna = sequence_string.upper().replace(\"N\" , \"\")\n",
        "    self.l = l_mer_length        # Default word length = 11\n",
        "    self.query = query\n",
        "    self.q = len(query)\n",
        "    self.dna_sequence = dna\n",
        "    self.n = len(dna)\n",
        "    self.score_params = { \"match\":1, \"gap_insertion\": -1 , \"mismatch\": -1 , \"gap_extension\": -1 }   \n",
        "    self.extension_threshold =  math.ceil(hssp_score*3/4) # If the score drops below 1 while extension we stop extending\n",
        "    self.E_threshold = 0.005\n",
        "    self.HSSP_score = hssp_score # high scoring sequence pairs score\n",
        "    self.nuc_dict = {\"A\":0, \"T\":3, \"C\":1 , \"G\":2, }  # Using numerical representation of nucleotides\n",
        "    self.sequence_words = {}     # { key_of_the_word: [list of starting positions] }\n",
        "    self.query_words = []        # List of query l-mers and their mutation with ( with score > HSSP Score )\n",
        "    self.matched = {}            # { starting_pos_in_query : [list of starting pos in DNA sequence ]}\n",
        "    self.local_alignments = []   # List of [ [ query_start_pos, query_end_pos, seq_start_pos, seq_end_pos]]\n",
        "    self.param_lambda = 0        # parameter for statistical significance\n",
        "    self.param_k  = 0\n",
        "    \n",
        "    # Lists for graphs \n",
        "    self.pos_list = []  #starting positions\n",
        "    self.E_list = []    #corresponding E value\n",
        "    self.prob_list = [] #corresponding probabilities\n",
        "\n",
        "    self.finding_sequence_words()\n",
        "    self.finding_query_words()\n",
        "    self.seeding()\n",
        "    self.extension()\n",
        "    self.calculate_params()\n",
        "    self.display()\n",
        "    self.end_time = time.time()\n",
        "\n",
        "  def calculate_params(self):\n",
        "    \"\"\"\n",
        "    Calculating constants lambda and K , for Karlin Atschul equation. These constants are dependent on the probability of nucleotide in query and DNA Sequence and \n",
        "    the scoring scheme. \n",
        "    \"\"\"\n",
        "    freq_nuc = {}\n",
        "    nucs=\"ATCG\"\n",
        "    for nuc in nucs:\n",
        "      freq_nuc[nuc]= [ self.query.count(nuc)/self.q , self.dna_sequence.count(nuc)/self.n]\n",
        "    q = 0 #negative S(i,j) \n",
        "    p = 0 #positive S(i,j)\n",
        "    for i in nucs:\n",
        "      p  += freq_nuc[i][0]*freq_nuc[i][1]\n",
        "      for j in nucs:\n",
        "        if j != i :\n",
        "          q  += freq_nuc[i][0]*freq_nuc[j][1]\n",
        "    \n",
        "    self.param_lambda  = math.log(p/q)\n",
        "    self.param_k = ((q-p)**2)/q\n",
        "    \n",
        "  \n",
        "  def probability(self, i):\n",
        "  \n",
        "    score = self.score_words(self.query[i[0]: i[1]+1] , self.dna_sequence[i[2]: i[3]+1 ], string=True )\n",
        "    E = self.param_k*self.q*self.n*math.exp(self.param_lambda*score)\n",
        "    prob = 1 - math.exp(-E)\n",
        "    return (prob, E , score)\n",
        "\n",
        "\n",
        "  def key(self, list_of_seq):\n",
        "    \"\"\"\n",
        "    Converts the string representation of a k-mer to encoded key using the numerical representation\n",
        "      # e.g. TGAC = 3*(4**3) +  2*(4**2) + 0*(4**0) + 1*(4**1)\n",
        "    \"\"\"\n",
        "    list_of_keys = []\n",
        "    for seq in list_of_seq:\n",
        "      key = 0 \n",
        "      for i in range(len(seq)):\n",
        "        key += self.nuc_dict[seq[i].upper()]*(4**(len(seq)-i-1))\n",
        "      list_of_keys.append(key)\n",
        "    return list_of_keys\n",
        "\n",
        "  def finding_sequence_words(self):\n",
        "    \"\"\"\n",
        "    Finding all the l-mers of the given DNA sequence and hashing their key with the starting position.\n",
        "    \"\"\"\n",
        "    \n",
        "    for i in range(self.n - self.l + 1 ):\n",
        "      key = self.key([self.dna_sequence[i:self.l+i]])[0]\n",
        "      if key in self.sequence_words:\n",
        "        self.sequence_words[key].append(i)\n",
        "      else:\n",
        "        self.sequence_words[key] = [i]\n",
        "\n",
        "  def finding_query_words(self):\n",
        "    \"\"\"\n",
        "    Finding the all the l-mers and their mutation with ( with score > HSSP Score ) of the given query strings\n",
        "    \"\"\"\n",
        "\n",
        "    for i in range(self.q - self.l + 1):  \n",
        "      sequences = self.find_mutations(self.query[ i : self.l+i ].upper())\n",
        "      self.query_words.append(self.key(list(set(sequences))))\n",
        "\n",
        "  def find_mutations(self, seq):\n",
        "    \"\"\"\n",
        "    Finds all the possible sequences with 0, 1, 2 mutations in the given sequence\n",
        "    \"\"\"\n",
        "    list_of_mutations = [seq]\n",
        "    nucleotide = [ 'A', 'T', 'G', 'C'] \n",
        "    for i in range(len(seq)):\n",
        "      for  t in nucleotide:\n",
        "        if(t != seq[i]):\n",
        "          mut = seq\n",
        "          mut = mut[:i] + t + mut[i+1:]\n",
        "          list_of_mutations.append(mut)\n",
        "          for j in range(i+1, len(seq)):\n",
        "            for t in nucleotide:\n",
        "              mut2 = mut\n",
        "              if(t != mut[j]):\n",
        "                mut2 = mut2[:j] + t + mut2[j+1:]\n",
        "                list_of_mutations.append(mut2)\n",
        "    return list_of_mutations\n",
        "\n",
        "  def seeding(self):\n",
        "    \"\"\"\n",
        "    Finding matches between sequence words and HSSP from query string \n",
        "    \"\"\"\n",
        "    for pos in range(len(self.query_words)):\n",
        "      for seq in self.query_words[pos]:\n",
        "        if seq in self.sequence_words:\n",
        "          if pos in self.matched:\n",
        "            self.matched[pos].extend(self.sequence_words[seq])\n",
        "          else:\n",
        "            self.matched[pos] = self.sequence_words[seq]\n",
        "  \n",
        "  def extension(self):\n",
        "    \"\"\"\n",
        "    Extending the seeded matches using Smith Waterman Algorithm ( ungapped )\n",
        "    \"\"\"\n",
        "    for query_pos in self.matched:\n",
        "      for seq_pos in self.matched[query_pos]:\n",
        "        score = self.score_words(query_pos, seq_pos)\n",
        "        i = query_pos\n",
        "        j = seq_pos\n",
        "        s = score\n",
        "        while ( i > 0 and j > 0 and s >= self.extension_threshold ):\n",
        "          if self.query[i-1] == self.dna_sequence[j-1]:\n",
        "            s+= self.score_params[\"match\"]\n",
        "          else:\n",
        "            s+= self.score_params[\"mismatch\"]\n",
        "          i-=1\n",
        "          j-=1\n",
        "        query_start = i\n",
        "        seq_start = j\n",
        "        i = query_pos + self.l - 1\n",
        "        j = seq_pos + self.l - 1\n",
        "        s = score\n",
        "        while ( i < (self.q-1) and j < (self.n -1) and s >= self.extension_threshold ):\n",
        "          if self.query[i+1] == self.dna_sequence[j+1]:\n",
        "            s+= self.score_params[\"match\"]\n",
        "          else:\n",
        "            s+= self.score_params[\"mismatch\"]\n",
        "          i+=1\n",
        "          j+=1\n",
        "        query_stop = i\n",
        "        seq_stop = j\n",
        "        alignment = [query_start, query_stop, seq_start, seq_stop]\n",
        "        if alignment not in self.local_alignments:\n",
        "          self.local_alignments.append(alignment)\n",
        "  \n",
        "  def score_words(self, query_pos, seq_pos, string=False):\n",
        "    \"\"\"\n",
        "    Return score for a pair of words\n",
        "    \"\"\"\n",
        "    score =  0\n",
        "    if string:\n",
        "      for i in range(len(query_pos)):\n",
        "        if query_pos[i] == seq_pos[i]:\n",
        "          score+= self.score_params[\"match\"]\n",
        "        # else:\n",
        "        #   score+= self.score_params[\"mismatch\"]\n",
        "      score = (score*100)/self.q\n",
        "    else:   \n",
        "      for i in range(self.l):\n",
        "        if query_pos+i >= self.q or seq_pos+i >= self.n:\n",
        "          break\n",
        "        elif self.query[query_pos+i] == self.dna_sequence[seq_pos+i]:\n",
        "          score+= self.score_params[\"match\"]\n",
        "        else:\n",
        "          score+= self.score_params[\"mismatch\"]\n",
        "    return score\n",
        "  \n",
        "  def display(self):\n",
        "    \"\"\"\n",
        "    Displays local alignment\n",
        "    \"\"\"\n",
        "    if len(self.local_alignments)==0 :\n",
        "      print ( \"No alignment found\")\n",
        "    count=0\n",
        "    for alignment in self.local_alignments:\n",
        "      prob, E , score  = self.probability(alignment)\n",
        "      self.pos_list.append(alignment[2])\n",
        "      self.E_list.append(E)\n",
        "      self.prob_list.append(prob)\n",
        "      if E<= self.E_threshold:\n",
        "        count=count+1\n",
        "        print(\"Probality:\", prob)\n",
        "        print(\"E Value:  \", E)\n",
        "        print(\"Score: \", score)\n",
        "        print(\"Query Cover: \", round((alignment[1]-alignment[0]+1)*100/self.q , 2), \"%\")\n",
        "        q_pos = alignment[0]\n",
        "        seq_pos = alignment[2]\n",
        "        s= \"DNA Sequence(Start Position=\"+ str(seq_pos)+\"): \" \n",
        "        k = len(s)*\" \"\n",
        "        ss = \"\\nQuery(Start Position=\"+ str(q_pos)+\"): \"\n",
        "        kk = (len(s)-len(ss))*\" \"\n",
        "        print( s + self.dna_sequence[alignment[2]: alignment[3]+1 ])\n",
        "        print(k, end =\"\")\n",
        "        a=0\n",
        "        \n",
        "        cnt = alignment[3] - alignment[2] \n",
        "        while ( a <= cnt):\n",
        "          if self.query[q_pos+a] == self.dna_sequence[seq_pos+a]:\n",
        "            print(\"|\" , end =\"\")\n",
        "          else: \n",
        "            print(\" \" , end =\"\")\n",
        "          a+=1\n",
        "\n",
        "        print( ss +kk+\" \"+ self.query[alignment[0]: alignment[1]+1 ], end=\"\\n\\n\\n\")\n",
        "    print(\"Total Number of Hits Found: \",count)   \n",
        "  \n",
        "        "
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "g0gDQsc9T39Q",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 697
        },
        "outputId": "6cb7ad8d-2bdf-4aa0-ad94-d43f8e2514d4"
      },
      "source": [
        "def main():\n",
        "  #Sequence string fasta\n",
        "  x=open(\"/content/ex.fasta\")\n",
        "  a=x.read()\n",
        "  ind=a.find(\"\\n\")\n",
        "  a=a[ind:]\n",
        "  seque=\"\"\n",
        "  for i in range(len(a)):\n",
        "    if a[i]!=\"\\n\":\n",
        "      seque=seque+str(a[i])\n",
        "  print(\"Printing length of Subject sequence: \", len(seque) )\n",
        "\n",
        "  #Query string fasta\n",
        "  x1=open(\"/content/ex2.fasta\")\n",
        "  a1=x1.read()\n",
        "  ind1=a1.find(\"\\n\")\n",
        "  a1=a1[ind1:]\n",
        "  quer=\"\"\n",
        "  for i in range(len(a1)):\n",
        "    if a1[i]!=\"\\n\":\n",
        "      quer=quer+str(a1[i])\n",
        "  print(\"Printing length of Query sequence: \", len(quer))\n",
        "\n",
        "  blast = BLAST( query_string = quer , sequence_string = seque )\n",
        "  print(\"Time Required: \",blast.end_time-blast.start_time)\n",
        "\n",
        "main()"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Printing length of Subject sequence 29903\n",
            "Printing length of Query sequence 198\n",
            "Probality: 0.002827025456615062\n",
            "E Value:   0.0028310290403503148\n",
            "Score:  21.27659574468085\n",
            "Query Cover:  40.96 %\n",
            "DNA Sequence(Start Position=11604): GCTAGTTTATTGTTTCTTAGGCTATTTTTGTACTTGTTACTTTGGCCTCTTTTGTTTACTCAACCGCTACTTTAGAC\n",
            "                                      |||||| || ||| ||    | ||||| |  || ||  |||    | |||| |||  |       |  |||    \n",
            "Query(Start Position=70):           AGTAGTTTCTTCTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTT\n",
            "\n",
            "\n",
            "Probality: 0.0010359593143238754\n",
            "E Value:   0.0010364962910637922\n",
            "Score:  22.340425531914892\n",
            "Query Cover:  43.09 %\n",
            "DNA Sequence(Start Position=9572): GTGTTTATTCTGTTATTTACTTGTACTTGACATTTTATCTTACTAATGATGTTTCTTTTTTAGCACATATTCAGTGGATGG\n",
            "                                     |||| |||| || |||  || |  ||    |||| | ||  |  |  | ||| ||||||      | ||   |   |  \n",
            "Query(Start Position=72):          TAGTTTCTTCTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTT\n",
            "\n",
            "\n",
            "Probality: 3.077818934549281e-05\n",
            "E Value:   3.0778663003719636e-05\n",
            "Score:  26.06382978723404\n",
            "Query Cover:  51.6 %\n",
            "DNA Sequence(Start Position=11059): CAATGGTCTTTGTTCTTTTTTTTGTATGAAAATGCCTTTTTACCTTTTGCTATGGGTATTATTGCTATGTCTGCTTTTGCAATGATGTTTGTCAAAC\n",
            "                                      | |   ||| ||||||||||| | |     |   |||||   ||||  | |   | || ||  | | | |  ||||    |  | ||| |     \n",
            "Query(Start Position=67):           ACAAGTAGTTTCTTCTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTT\n",
            "\n",
            "\n",
            "Probality: 1.862333350088896e-05\n",
            "E Value:   1.8623506917355346e-05\n",
            "Score:  26.595744680851062\n",
            "Query Cover:  51.6 %\n",
            "DNA Sequence(Start Position=11063): GGTCTTTGTTCTTTTTTTTGTATGAAAATGCCTTTTTACCTTTTGCTATGGGTATTATTGCTATGTCTGCTTTTGCAATGATGTTTGTCAAACATAA\n",
            "                                      ||||| || |||||||| | |     |   |||||   ||||  | |   | || ||  | | | |  ||||    |  | ||| |  | |    \n",
            "Query(Start Position=78):           CTTCTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTGACCCGGG\n",
            "\n",
            "\n",
            "Total Number of Hits Found:  4\n",
            "Time Required:  12.855626821517944\n"
          ],
          "name": "stdout"
        }
      ]
    }
  ]
}